[REGRESSION]: import with TypeScript path not resolved correctly since v1.35.0

**Context:**
- GOOD Playwright Version: 1.34.3
- BAD Playwright Version: 1.35.0 and above
- Operating System: macOS

**Code Snippet**

Full reduced test case: https://github.com/OliverJAsh/playwright-pkg-require-with-path

Contents inlined below:

`tsconfig.json`:
```json
{
    "compilerOptions": {
        "baseUrl": ".",
        "paths": {
            "app/*": ["app/*"]
        }
    }
}

```

`app/my-pkg/main.ts`:
```ts
export const foo = 1;

```

`app/my-pkg/package.json`:
```json
{
  "main": "main.ts"
}

```

`tests/example.spec.ts`:
```ts
// ❌ Using path, it doesn't work.
import * as x from 'app/my-pkg';
// ✅ Not using path, it works.
// import * as x from '../app/my-pkg';

console.log(x);
```

**Describe the bug**

Starting with Playwright 1.35.0, Playwright returns an error: `Error: Cannot find module 'app/my-pkg'`.

Note that if we change the `import` to avoid using the TS path (`app/*`), it works:

```ts
import * as x from '../app/my-pkg';

console.log(x);
``` 

```sh-session
$ yarn
$ yarn playwright test
yarn run v1.22.19
Error: Cannot find module 'app/my-pkg'
Require stack:
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/tests/example.spec.ts
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/transform/transform.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/common/config.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/reporters/json.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/reporters/raw.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/reporters/html.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/runner/reporters.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/runner/runner.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/lib/cli.js
- /Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/node_modules/@playwright/test/cli.js

   at example.spec.ts:2

  1 | // ❌ Using path, it doesn't work.
> 2 | import * as x from 'app/my-pkg';
    | ^
  3 | // ✅ Not using path, it works.
  4 | // import * as x from '../app/my-pkg';
  5 |

    at Object.<anonymous> (/Users/oliver/Code/Reduced test cases/playwright-pkg-require-with-path/tests/example.spec.ts:2:1)

Error: No tests found







To open last HTML report run:

  npx playwright show-report

error Command failed with exit code 1.
```

--- a/packages/playwright-test/src/util.ts
+++ b/packages/playwright-test/src/util.ts
@@ -342,8 +342,13 @@ export function resolveImportSpecifierExtension(resolved: string): string | unde
     }
     break;  // Do not try '' when a more specific extesion like '.jsx' matched.
   }
-  // try directory imports last
+
   if (dirExists(resolved)) {
+    // If we import a package, let Node.js figure out the correct import based on package.json.
+    if (fileExists(path.join(resolved, 'package.json')))
+      return resolved;
+
+    // Otherwise, try to find a corresponding index file.
     const dirImport = path.join(resolved, 'index');
     return resolveImportSpecifierExtension(dirImport);
   }

--- a/tests/playwright-test/resolver.spec.ts
+++ b/tests/playwright-test/resolver.spec.ts
@@ -505,3 +505,40 @@ test('should support extends in tsconfig.json', async ({ runInlineTest }) => {
   expect(result.passed).toBe(1);
   expect(result.exitCode).toBe(0);
 });
+
+test('should import packages with non-index main script through path resolver', async ({ runInlineTest }) => {
+  const result = await runInlineTest({
+    'app/pkg/main.ts': `
+      export const foo = 42;
+    `,
+    'app/pkg/package.json': `
+      { "main": "main.ts" }
+    `,
+    'package.json': `
+      { "name": "example-project" }
+    `,
+    'playwright.config.ts': `
+      export default {};
+    `,
+    'tsconfig.json': `{
+      "compilerOptions": {
+        "baseUrl": ".",
+        "paths": {
+          "app/*": ["app/*"],
+        },
+      },
+    }`,
+    'example.spec.ts': `
+      import { foo } from 'app/pkg';
+      import { test, expect } from '@playwright/test';
+      test('test', ({}) => {
+        console.log('foo=' + foo);
+      });
+    `,
+  });
+
+  expect(result.exitCode).toBe(0);
+  expect(result.passed).toBe(1);
+  expect(result.output).not.toContain(`find module`);
+  expect(result.output).toContain(`foo=42`);
+});

